target-chains/near/receiver/src/governance.rs

```diff
@@ -265,26 +321,27 @@ impl Pyth {
     ) -> Result<(), Error> {
         use GovernanceAction::*;
 
-        if !is_promise_success() {
-            return Err(Error::VaaVerificationFailed);
-        }
+        ensure!(is_promise_success(), VaaVerificationFailed);
 
         // Get Storage Usage before execution.
         let storage = env::storage_usage();
 
         // Deserialize VAA, note that we already deserialized and verified the VAA in `process_vaa`
         // at this point so we only care about the `rest` component which contains bytes we can
         // deserialize into an Action.
-        let vaa = hex::decode(&vaa).unwrap();
+        let vaa = hex::decode(vaa).map_err(|_| InvalidPayload)?;
         let (_, rest): (wormhole::Vaa<()>, _) =
-            serde_wormhole::from_slice_with_payload(&vaa).map_err(|_| Error::InvalidPayload)?;
-
-        match GovernanceAction::deserialize(rest)? {
-            ContractUpgrade(codehash) => self.set_upgrade_hash(codehash),
-            SetDataSources(sources) => self.set_sources(sources),
-            SetGovernanceSource(source) => self.set_gov_source(source),
-            SetStalePriceThreshold(threshold) => self.set_stale_price_threshold(threshold),
-            SetUpdateFee(fee) => self.set_update_fee(fee),
+            serde_wormhole::from_slice_with_payload(&vaa).map_err(|_| InvalidPayload)?;
+
+        match GovernanceInstruction::deserialize(rest)?.action {
+            UpgradeContract { codehash } => self.set_upgrade_hash(codehash),
```

Comment:
Since the payload of contract upgrade actions are chain-specific, it might be worth adding a check that the action is destined only for NEAR.

Reply:
You're right, this check is performed later on but that's a bad place for it, I will lift it up to here.